### PR TITLE
Make Exit Conditional and change RA capture Production Lock back to Exit Lock.

### PR DESCRIPTION
**Pull request by MustaphaTR**
I added DisableExitOnCondition to Production as i couldn't get making Exit conditional work, but i guess that would be the correct way to do it. I don't wanna mess with it again rn, but someone else can do it, this hack can be removed. Current changes should do the job for now.

### CODE REVIEW

**Comment by pchote** [comment]
Do you have a branch with your attempt at making Exit conditional?

**Comment by MustaphaTR** [comment]
No, i think i lost that code. I'm planning to reattempt in soon tbh, but not sure when i can find time for it.

**Comment by MustaphaTR** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/ExternalCapturable.cs [hunk 1]
```diff
@@ -48,33 +52,39 @@ public bool CanBeTargetedBy(Actor captor, Player owner)
...
 
 			Captor = captor;
 		}
 
 		public void EndCapture()
 		{
-			var building = self.TraitOrDefault<Building>();
-			if (building != null)
-				building.Unlock();
+			if (token != ConditionManager.InvalidConditionToken)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/ExternalCapturable.cs]: This probably wants a `null` check as well.

**Comment by pchote** [comment]
Removing the building lock simplifies my fixes for #15076 and #15123, so adding the dependency tag.  If the delay isn't going to hurt modding too badly then I would prefer that we defer this until after the next playtest is branched so that all of the Captures changes can be done in the one cycle.

**Comment by MustaphaTR** [comment]
I think we want that soon for Shattered Paradise, @ABrandau ?

**Comment by GraionDilach** [comment]
This can wait for that IMO. The locomotor shift is already a major change inbetween.

**Comment by pchote** [comment]
The first commit should be safe enough, but the latter two require update rules and manual changes that overlap strongly with the building lock removal.  Merging that will introduce a blocker to remove the rest of the locking code because we really don't want to drag the manual rule changes for that over multiple releases.

**Comment by ABrandau** [comment]
Yeah, the removal of the engineer lock would be very nice to have, but we can wait, np.

**Comment by pchote** [comment]
I'd like to get #15661 in first, and then this one ASAP afterwards, before #15045.

**Comment by pchote** [comment]
Can you please rebase this now?

**Comment by MustaphaTR** [comment]
Rebased.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 2]
```diff
@@ -89,8 +89,8 @@ public override Activity Tick(Actor self)
 				}
 			}
 
-			var exit = dest.Info.FirstExitOrDefault(null);
-			var offset = (exit != null) ? exit.SpawnOffset : WVec.Zero;
+			var exit = dest.FirstExitOrDefault(null);
+			var offset = (exit != null) ? exit.Info.SpawnOffset : WVec.Zero;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: The parenthesis are unnecessary (same for the ones in `Aircraft.cs`). Not a hard blocker, though.

**Review by abcdefg30 — DISMISSED** [review]

**Comment by pchote** [comment]
I pushed over this with fixes for @abcdefg30's style nit, and a couple more of my own (`RequiresCondition` should be listed first, factored out disabled trait checks instead of making the linq queries even messier).

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/72eb6e8e950b08d3bfb01af502bcb460a70b5759

**Pull request merged**